[BB2-896] Responsive Design NodeJS React Sample

JFU-GIT opened the pull request:
<!--
You've got a Pull Request you want to submit? Awesome!
This PR template is here to help ensure you're setup for success:
  please fill it out to help ensure that your PR is complete and ready for approval.
-->

**JIRA Ticket:**
[BB2-896](https://jira.cms.gov/browse/BB2-896)

**User Story or Bug Summary:**
<!-- Please copy-paste the brief user story or bug description that this PR is intended to address. -->
As a 3rd party app, I should be able to understand what to do when I receive certain error messages from BB2 so that I can properly respond to my own users

AC:

Draft documentation for review for:

- What to do when receiving a 403 error
- What to do when receiving a 404 error
- What to do when receiving a 500 error
- how to implement exponential backoff

In the Node.JS and Python sample apps add in responsive design to show:

- What to do when receiving a 403 error
- What to do when receiving a 404 error
- What to do when receiving a 500 error
- how to implement exponential backoff

### What Does This PR Do?

1. Create a document [BB2 API Error Responses & Handling](https://confluence.cms.gov/pages/viewpage.action?pageId=540087166) where Blue Button 2.0 API Error Responses (HTTP status codes) and Recommended Error Handling Are Documented, including, among other things, how to retry request upon certain error responses.
2. Add improved error reporting to the sample app where it is applicable, and implement retry logic (exponential back off) when applicable.

Extra:

Avoid tweaking of the code for native OS run or in Docker run, introduce a shell script to start the sample in native OS or in Docker, make client and server code context aware such that the AUTH2 authentication flow functions when the sample is running in docker or on native OS.

<!--
Add detailed description & discussion of changes here.
The contents of this section should be used as your commit message (unless you merge the PR via a merge commit, of course).

Please follow standard Git commit message guidelines:
* First line should be a capitalized, short (50 chars or less) summary.
* The rest of the message should be in standard Markdown format, wrapped to 72 characters.
* Describe your changes in imperative mood, e.g. "make xyzzy do frotz" instead of "[This patch] makes xyzzy do frotz" or "[I] changed xyzzy to do frotz", as if you are giving orders to the codebase to change its behavior.
* List all relevant Jira issue keys, one per line at the end of the message, per: <https://confluence.atlassian.com/jirasoftwarecloud/processing-issues-with-smart-commits-788960027.html>.

Reference: <https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project>.
-->


### What Should Reviewers Watch For?

* Are error responses from API calls handled as instructed by the document?
* Are any additional documentation updates needed?
* Are there any unhandled and/or untested edge cases you can think of?
* Is the retry work as expected?

If you're reviewing this PR, please check these things, in particular:

* Review the document on API error responses
* Verify the retry logic works (exponential back off)

To facilitate the verification of retry failed requests, an instrumented BB2 server is deployed to sandbox canary:
put below line into your host file on you local machine:

10.246.44.56	sandbox.bluebutton.cms.gov
(new canary deployed on Jan 19, 2022)

then:

1. create an app on sandbox (or using an existing one)
2. make sure the url is among the redirect urls: http://server:3001/api/bluebutton/callback/
3. copy the client_id, client_secret to the sample app server config.ts
4. start sample app in docker: go to local repo base and run ./start.sh docker

Wait until the server and client components are up and running.
note, a waffle switch "simulate_500_error" is provisioned on sandbox which when "active", will produce a 500 error on fhir request, with error message: "Service temporarily unavailable, try again later, network connection refused".

From client webUI, click button "Authorize" to complete the beneficiary medicare.gov login, and grant access to the app, the corresponding EOB search will hit the simulated 500 network connection error, and the failed EOB request met the criteria for retry.

below is a sample server stdout showing the retry process, the 3rd attempt succeeded (by setting switch "simulate_500_error" to false soon after 2nd retry ...)

`server_1  | [2021-11-19T00:16:56.149Z] INFO: [object Object]
server_1  | Error message: [ Request failed with status code 500 ]
server_1  | response code: 500
server_1  | response text: [object Object]
server_1  | Request failed and is retryable, entering retry process...
server_1  | wait  5  seconds...
server_1  | retry attempts:  1
server_1  | retry error: [ Request failed with status code 500 ]
server_1  | response code:  500
server_1  | response data:  {
server_1  |   detail: 'Service temporarily unavailable, try again later, network connection refused.
'
server_1  | }
server_1  | wait  10  seconds...
server_1  | retry attempts:  2
server_1  | retry error: [ Request failed with status code 500 ]
server_1  | response code:  500
server_1  | response data:  {
server_1  |   detail: 'Service temporarily unavailable, try again later, network connection refused.
'
server_1  | }
server_1  | wait  20  seconds...
server_1  | retry attempts:  3
server_1  | retry successful:
server_1  | {
server_1  |   resourceType: 'Bundle',
server_1  |   id: '787094eb-dca9-4af4-9b39-87e040f65417',
server_1  |   meta: { lastUpdated: '2021-09-23T13:36:42.530-04:00' },
server_1  |   type: 'searchset',
server_1  |   total: 51,
server_1  |   link: [
server_1  |     {
server_1  |       relation: 'first',
server_1  |       url: 'https://sandbox.bluebutton.cms.gov/v2/fhir/ExplanationOfBenefit?_format=appl
ication%2Fjson%2Bfhir&startIndex=0&_count=10&patient=-20000000000002'
server_1  |     },
server_1  |     {
server_1  |       relation: 'next',
server_1  |       url: 'https://sandbox.bluebutton.cms.gov/v2/fhir/ExplanationOfBenefit?_format=appl
ication%2Fjson%2Bfhir&startIndex=10&_count=10&patient=-20000000000002'
server_1  |     },
server_1  |     {
server_1  |       relation: 'last',
server_1  |       url: 'https://sandbox.bluebutton.cms.gov/v2/fhir/ExplanationOfBenefit?_format=appl
ication%2Fjson%2Bfhir&startIndex=50&_count=10&patient=-20000000000002'
server_1  |     },
server_1  |     {
server_1  |       relation: 'self',
server_1  |       url: 'https://sandbox.bluebutton.cms.gov/v2/fhir/ExplanationOfBenefit/?_count=10&_
format=application%2Fjson%2Bfhir&patient=-20000000000002&startIndex=0'
server_1  |     }
server_1  |   ],
server_1  |   entry: [
server_1  |     { resource: [Object] },
server_1  |     { resource: [Object] },
server_1  |     { resource: [Object] },
server_1  |     { resource: [Object] },
server_1  |     { resource: [Object] },
server_1  |     { resource: [Object] },
server_1  |     { resource: [Object] },
server_1  |     { resource: [Object] },
server_1  |     { resource: [Object] },
server_1  |     { resource: [Object] }
server_1  |   ]
server_1  | }
server_1  | config:
server_1  | {
server_1  |   url: 'https://sandbox.bluebutton.cms.gov/v2/fhir/ExplanationOfBenefit/',
server_1  |   method: 'get',
server_1  |   headers: {
server_1  |     Accept: 'application/json, text/plain, */*',
server_1  |     Authorization: 'Bearer k8O6XDxBf2RCLmD0xzAiGKlWvnOcRS',
server_1  |     'User-Agent': 'axios/0.21.1'
server_1  |   },
server_1  |   params: {
server_1  |     code: 'mRpSQkterTO9xzIfXEab6hRCun4ka0',
server_1  |     state: 'k6hb1zQg_TnBOsmozJVEyDwFxxLHd2tJrD8Q39WPjzE'
server_1  |   },
server_1  |   transformRequest: [ [Function: transformRequest] ],
server_1  |   transformResponse: [ [Function: transformResponse] ],
server_1  |   timeout: 0,
server_1  |   adapter: [Function: httpAdapter],
server_1  |   xsrfCookieName: 'XSRF-TOKEN',
server_1  |   xsrfHeaderName: 'X-XSRF-TOKEN',
server_1  |   maxContentLength: -1,
server_1  |   maxBodyLength: -1,
server_1  |   validateStatus: [Function: validateStatus],
server_1  |   data: undefined
server_1  | }
`




### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then here's a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence: N/A.
    * Does this PR add any new software dependencies? **No**.
    * Does this PR modify or invalidate any of our security controls? **No**.
    * Does this PR store or transmit data that was not stored or transmitted before? **No**.
* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR should not be merged unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons? **No**.


### What Needs to Be Merged and Deployed Before this PR?

<!--
Add some items to the following list, or remove the entire section if it doesn't apply.

Common items include:
* Database migrations (which should always be deployed by themselves, to reduce risk).
* New features in external dependencies (e.g. BFD).
-->

This PR cannot be either merged or deployed until the following pre-requisite changes have been fully deployed:

* CMSgov/some_repo#42
  
### Any Migrations?

<!--
Make sure to work with whoever is doing the deploy so they are aware of any migrations that may need to be run
-->

* [ ] Yes, there are migrations
  * [ ] The migrations should be run PRIOR to the code being deployed
  * [ ] The migrations should be run AFTER the code is deployed
  * [ ] There is a more complicated migration plan (downtime, etc) <!-- Make sure to include the details of the plan below -->
* [x] No migrations


### Submitter Checklist

<!--
Helpful hint: if needed, Git allows you to edit your PR's commits and history, prior to merge.
See these resources for more information:

* <https://dev.to/maxwell_dev/the-git-rebase-introduction-i-wish-id-had>
* <https://raphaelfabeni.com/git-editing-commits-part-1/>
-->

I have gone through and verified that...:

* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] I have named this PR and its branch such that they'll be automatically be linked to the (most) relevant Jira issue, per: <https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html>.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. BFD, SLS, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.
… docker or natively.

nbragdon commented:
@JFU-GIT We really shouldn't be creating new startup scripts. I think the idea of using docker was that it was platform agnostic. Now adding in the shell scripts makes it less so. I think we are just going to have to work with what is there, or somehow integrate it into the docker setup. I think this PR shouldn't change any of that though.

JFU-GIT commented:
> @JFU-GIT We really shouldn't be creating new startup scripts. I think the idea of using docker was that it was platform agnostic. Now adding in the shell scripts makes it less so. I think we are just going to have to work with what is there, or somehow integrate it into the docker setup. I think this PR shouldn't change any of that though.

understand the point, but sometime start script reduce the manual steps...

nbragdon commented on server/src/utils/bb2.ts:
```diff
@@ -33,7 +33,8 @@ export async function getAccessToken(code: string, state: string | undefined) {
 
     const BB2_ACCESS_TOKEN_URL = envConfig.bb2BaseUrl + '/' + db.settings.version + '/o/token/';
-
+    // const BB2_ACCESS_TOKEN_URL = 'http://host.docker.internal:8000/' + db.settings.version + '/o/token/';
```
Lets remove this commented out line

nbragdon commented on server/src/utils/bb2.ts:
```diff
@@ -1,9 +1,9 @@
 import db from './db';
 import config from '../configs/config';
 import { generateCodeChallenge, generateRandomState } from './generatePKCE';
+import { endpointPost } from './call'
```
Can we rename this file to something more descriptive than `call`?

nbragdon reviewed (commented):
@JFU-GIT I don't think this cron/queue approach is really ideal for exponential backoff... I think we should do something more like in this article: https://advancedweb.hu/how-to-implement-an-exponential-backoff-retry-strategy-in-javascript/

Specifically the solution listed under: Rejection-based retrying for async/await syntax

nbragdon commented on server/src/utils/call.ts:
```diff
@@ -0,0 +1,66 @@
+        headers: extra}, true);
+}
+
+export async function endpointGet(endpoint_url: string, req_qry: any, token: string) {
```
I'm thinking something more like... `getFhirResource` might be a better naming schema here

nbragdon replied:
I think we only would want to do retries when requesting fhir data, I don't think we want to introduce them into the auth flow

JFU-GIT replied:
it's a unified code so we have a centralized place to add control, in case something else needed, it can be added here conveniently.

nbragdon commented on README.md:
```diff
@@ -57,3 +57,33 @@ Read the DEVELOPER NOTES found in the code to understand the application
 and where you will need to make adjustments/changes as well as some 
 suggestions for best practices.
 
+Usage Examples
```
Lets keep the startup instructions as they were

nbragdon commented on docker-compose.yml:
```diff
@@ -7,6 +7,7 @@ services:
       dockerfile: ./Dockerfile
     ports: 
       - "3001:3001"
+      - "9229:9229"
```
why was the new port added?

JFU-GIT replied:
for remote debugging

dtisza1 commented on server/src/index.ts:
```diff
@@ -1,7 +1,11 @@
 import logger from '@shared/Logger';
+import { schedule } from './utils/retry';
 
+// start retry cron job
+logger.info('Schedule retry cron job...');
+schedule();
```
> @JFU-GIT I don't think this cron/queue approach is really ideal for exponential backoff... I think we should do something more like in this article: https://advancedweb.hu/how-to-implement-an-exponential-backoff-retry-strategy-in-javascript/
> 
> Specifically the solution listed under: Rejection-based retrying for async/await syntax

@JFU-GIT @nbragdon I agree that the use of node-cron scheduler and queuing seems complicated for the purpose.

I just read through the article and agree that the "rejection-based retrying" method seems like it is simpler and a more typical/standard way to implement this.  @nbragdon Have you used this type of method successfully in the past?

JFU-GIT replied:
a in place retry added.

dtisza1 reviewed (changes requested):
@JFU-GIT Just a few review comments. TY!

JFU-GIT commented on start.sh:
```diff
@@ -0,0 +1,95 @@
+#!/bin/bash
+
+#
+# Start the sample application:
+# 1) in docker or
+# 2) in native OS - type CTRL C to terminate both the client and server components of the sample app
+#
```
no problem to remove the start.sh script, the sample app is a node js react app and developer playing with it mostly know how to fire it up either in docker or natively, just add help info in readme seems sufficient.

dtisza1 replied:
@JFU-GIT That sounds great and the added help info would be nice to have.

nbragdon commented on client/package.json:
```diff
@@ -23,7 +23,8 @@
   "scripts": {
-    "start": "react-scripts start",
+    "start": "REACT_APP_CTX=docker react-scripts start",
+    "start2": "REACT_APP_CTX=native react-scripts start",
```
Let's rename this to something more descriptive. `start2` isn't really telling me how/why it's different. Something like `start-native` would be more descriptive.

nbragdon commented on client/src/setupProxy.js:
```diff
@@ -4,7 +4,7 @@ module.exports = function(app) {
         '/api',
         createProxyMiddleware({
-            target: 'http://server:3001',
+            target: (process.env.REACT_APP_CTX === 'docker' ? 'http://server:3001' : 'http://localhost:3001'),
```
Nice change 👍

nbragdon commented on client/yarn.lock:
```diff
@@ -3224,9 +3224,9 @@ caniuse-api@^3.0.0:
-  resolved "https://registry.yarnpkg.com/caniuse-lite/-/caniuse-lite-1.0.30001191.tgz#bacb432b6701f690c8c5f7c680166b9a9f0843d9"
-  integrity sha512-xJJqzyd+7GCJXkcoBiQ1GuxEiOBCLQ0aVW9HMekifZsAVGdj5eJ4mFB9fEhSHipq9IOk/QXFJUiIr9lZT+EsGw==
+  version "1.0.30001279"
+  resolved "https://registry.npmjs.org/caniuse-lite/-/caniuse-lite-1.0.30001279.tgz"
```
Looks like the registry changed from yarn to npm... any idea why this change was made?

JFU-GIT replied:
probably used below to update the browser db up to date to silence the warnings:
npx browserslist@latest --update-db

not sure if yarn upgrade works?

yarn upgrade caniuse-lite browserlist

nbragdon commented on server/package.json:
```diff
@@ -70,9 +70,11 @@
     "module-alias": "^2.2.2",
-    "morgan": "^1.10.0"
+    "morgan": "^1.10.0",
+    "node-cron": "^3.0.0"
```
Let's remove this package since it's no longer used

nbragdon commented on server/package.json:
```diff
@@ -70,9 +70,11 @@
+    "node-cron": "^3.0.0"
   },
   "devDependencies": {
+    "@types/node-cron": "^3.0.0",
```
Can also remove the typing

nbragdon commented on server/src/routes/Authorize.ts:
```diff
@@ -52,6 +52,7 @@ export async function authorizationCallback(req: Request, res: Response) {
         * to display or store the error
          */
         console.log(e);
+        logger.info(e)
```
If we want to start using the logger, we should probably move away from the console logs. Also in this case it should be `logger.error(e)`.

nbragdon commented on server/src/routes/Data.ts:
```diff
@@ -15,15 +15,9 @@ import { getLoggedInUser } from 'src/utils/user';
-    });
-    return response.data;
+    // get EOB end point
+    const response = await get(envConfig.bb2BaseUrl + '/' + db.settings.version + '/fhir/ExplanationOfBenefit/', req.query, `${loggedInUser.authToken?.access_token}`);
```
```suggestion
    const response = await get(`${envConfig.bb2BaseUrl}/${db.settings.version}/fhir/ExplanationOfBenefit/`, req.query, `${loggedInUser.authToken?.access_token}`);
```

nbragdon commented on server/src/routes/Data.ts:
```diff
@@ -42,45 +36,24 @@ export async function getBenefitDataEndPoint(req: Request, res: Response) {
-    });
-
+    // get Patient end point
+    const response = await get(envConfig.bb2BaseUrl + '/' + db.settings.version + '/fhir/Patient/', req.query, `${loggedInUser.authToken?.access_token}`);
```
```suggestion
    const response = await get(`${envConfig.bb2BaseUrl}/${db.settings.version}/fhir/Patient/`, req.query, `${loggedInUser.authToken?.access_token}`);
```

nbragdon commented on server/src/routes/Data.ts:
```diff
@@ -42,45 +36,24 @@ export async function getBenefitDataEndPoint(req: Request, res: Response) {
-    });
-
+    // get Coverage end point
+    const response = await get(envConfig.bb2BaseUrl + '/' + db.settings.version + '/fhir/Coverage/', req.query, `${loggedInUser.authToken?.access_token}`);
```
```suggestion
    const response = await get(`${envConfig.bb2BaseUrl}/${db.settings.version}/fhir/Coverage/`, req.query, `${loggedInUser.authToken?.access_token}`);
```

nbragdon commented on server/src/routes/Data.ts:
```diff
@@ -42,45 +36,24 @@ export async function getBenefitDataEndPoint(req: Request, res: Response) {
-    });
-
+    // get usrinfo end point
+    const response = await get(envConfig.bb2BaseUrl + '/' + db.settings.version + '/connect/userinfo', req.query, `${loggedInUser.authToken?.access_token}`);
```
```suggestion
    const response = await get(`${envConfig.bb2BaseUrl}/${db.settings.version}/connect/userinfo`, req.query, `${loggedInUser.authToken?.access_token}`);
```

nbragdon commented on server/src/utils/request.ts:
```diff
@@ -0,0 +1,101 @@
+import FormData from 'form-data';
+import logger from '@shared/Logger';
+
+export async function post(endpoint_url: string, data: FormData, extra: any) {
```
```suggestion
export async function post(endpoint_url: string, data: FormData, headers: any) {

```
We should avoid using the `any` typing whenever possible.. not sure if the headers have a generic typing or not. If so you should be able to get it from the response type of this function: `form.getHeaders()`

nbragdon replied:
Can you switch from snake -> camel?

nbragdon commented on server/src/utils/request.ts:
```diff
@@ -0,0 +1,101 @@
+        headers: extra}, true);
+}
+
+export async function get(endpoint_url: string, req_qry: any, token: string) {
```
```suggestion
export async function get(endpointUrl: string, params: any, authToken: string) {

```

nbragdon commented on server/src/utils/request.ts:
```diff
@@ -0,0 +1,101 @@
+        }}, true);
+}
+
+export async function request(config: any, retry_flag: boolean) {
```
```suggestion
export async function request(config: any, retryFlag: boolean) {

```

nbragdon commented on server/src/utils/request.ts:
```diff
@@ -0,0 +1,101 @@
+    return resp    
+}
+
+function is_retryable(error: any) {
```
```suggestion
function isRetryable(error: Error) {

```

nbragdon replied:
The convention for JS is to use camel case, not snake case. Kind of weird to have mixed cases throughout the code base.

nbragdon commented on server/src/utils/request.ts:
```diff
@@ -0,0 +1,101 @@
+    var resp = null
+    try {
+        resp = await axios(config);
+    } catch (error: any) {
```
```suggestion
    } catch (error: Error) {

```

nbragdon commented on server/src/utils/request.ts:
```diff
@@ -0,0 +1,101 @@
+}
+
+export async function request(config: any, retry_flag: boolean) {
+    var resp = null
```
```suggestion
    let resp = null

```
In modern JS it's common to use either `let` if the variable will change over time or `const` if it will never change.

nbragdon commented on server/src/utils/request.ts:
```diff
@@ -0,0 +1,101 @@
+        console.log("retry attempts: ", i+1)
+        try {
+            resp = await axios(config);
+            console.log("retry successful:")
```
Can we clean up some of these console logs?

JFU-GIT replied:
those are there to help reviewers verify the retry is working..., otherwise, it'd be more involved - e.g. do source line debugging...

dtisza1 replied:
@JFU-GIT Should the use of `console.log()` be converted to use the '@shared/Logger' instead here? And other locations for consistency across the code base?

nbragdon commented on server/src/utils/request.ts:
```diff
@@ -0,0 +1,101 @@
+    return resp
+}
+
+function sleep(ms: any) {
```
```suggestion
function sleep(time: number) {

```

nbragdon commented on ErrorResponses.md:
```diff
@@ -0,0 +1,42 @@
+### Authorization Requests
+
+
+| Status Code    | End Point URL   | Error Message              | Action            | Comments                                               |
```
I'm trying to decide if this is just too much information. Should we just highlight which type of requests they could retry on? Everything else feels a bit too much for a sample app. Basically it's a lot of text that essentially sums up to: "Don't do anything".

JFU-GIT replied:
though the purpose is to give more bb2 specifics, since HTTP response codes are already defined.

Can reduce the verbose-ness a bit.

nbragdon replied:
Want to take an attempt at trimming it down? Really focus on what they would need to know in order to successfully build an application against BB2.

JFU-GIT replied:
those are frequently used error responses among the full set http status codes, and it contains bb2 specific info, so there are added values, or let me know what should be trimmed? don't mind remove the whole mark down - remember it was added based on some feedback-:

nbragdon replied:
My only concern is that this table is a lot of text. I feel like most people wont read it all, and then they are missing out on important information. Were we to trim it down to just when they need to do something (ie retry) then they might catch it. If you feel strongly we can leave it though. Just clean up the typing issues I pointed out and we should be able to get this approved 👍

JFU-GIT replied:
for the enforce of typing, for those places that the type of the var or param is obvious, explicit type is given, but some places e.g. the err var of axios try catch, a single type could not cover all situations, e.g. the error thrown from axios end point call could be AxiosError but it could be something else, there are discussion on this from internet..., also ran across some people's code to handle the multiple error types - but it will pull in more code, may be we could log a backlog for "making node js react sample strong typed" as future improvement? thanks for brought this up.

nbragdon commented on ErrorResponses.md:
```diff
@@ -0,0 +1,41 @@
+
+## Overview
+
+This document serves as a supplementary to Blue Button 2.0 API Developer Documents, it gives more details on most commonly error responses and recommended error handling on the client side, e.g. guide lines on retrying the requests on certain errors.
```
```suggestion
This document serves as a supplementary to Blue Button 2.0 API Developer Documents. It gives more details on the most common error responses and how to properly handle them.

```

nbragdon commented on ErrorResponses.md:
```diff
@@ -0,0 +1,41 @@
+
+Auto retrying on FHIR Data read/search to overcome a FHIR backend network temporary downtime is recommended, but with sensible retry settings.
+
+Due to the involvement of end user (beneficiary), auto retrying on requests in an authorization flow does not seem practical, hence not recommended.
```
```suggestion
Due to the involvement of the end user (beneficiary), auto retrying requests in the authorization flow are not recommended.

```

nbragdon commented on ErrorResponses.md:
```diff
@@ -0,0 +1,41 @@
+
+App can choose to retry the requests on certain 500 errors heuristically depending on the root cause as indicated by the error messages, e.g. backend server is unavailable due to temporarily network down etc., the app can choose from various retry patterns that best fits the use case (e.g. exponential back off).
+
+Auto retrying on FHIR Data read/search to overcome a FHIR backend network temporary downtime is recommended, but with sensible retry settings.
```
```suggestion
Auto retrying (with sensible retry settings) on FHIR Data read/search to overcome a FHIR backend network temporary downtime is recommended.

```

nbragdon commented on ErrorResponses.md:
```diff
@@ -0,0 +1,41 @@
+
+### Retry
+
+App can choose to retry the requests on certain 500 errors heuristically depending on the root cause as indicated by the error messages, e.g. backend server is unavailable due to temporarily network down etc., the app can choose from various retry patterns that best fits the use case (e.g. exponential back off).
```
```suggestion


```

nbragdon commented:
@JFU-GIT Reminder that this also has to be done for the python app as well

nbragdon commented on client/src/components/records.tsx:
```diff
@@ -31,49 +37,79 @@ export default function Records({ }) {
-                    }
-                });
-                setRecords(records);
+                if (eobData.hasOwnProperty('entry')) {
```
```suggestion
                if (eobData.entry) {
```

JFU-GIT commented:
@nbragdon @dtisza1 rebased and retested plus some final touches on error handling - server logs error with code + details, while client got generic message on failed auth or failed fhir query.

dtisza1 reviewed (changes requested):
@JFU-GIT  This is looking good to me! Just having some possible changes and comments below:

* ErrorResponses.md:  This looks good in table format and split out from the Readme.md.  I think it would be good to put a link in the Readme.md that references this info.
* Readme.md:  Looks good to me. TY for adding the native startup! Just noticed it could use some cleanup with some sentences. For example, some sentences could begin capitalized.
* Should the console type logging be replaced in all areas with the '@shared/Logger'?
* Should the code be edited to use Camel case? Especially for the new code added in this PR. Could address previous code separately.
* The retry logic looks good to me! I didn't test it, but from your testing notes and server logs it looks to work as expected.

Thank you!

JFU-GIT commented:
@nbragdon @dtisza1 addressed feedbacks in the last commit, also deployed a sbx canary with instrumented BB2:

put below in hosts if want to test retry:

10.246.44.56 sandbox.bluebutton.cms.gov

dtisza1 reviewed (approved):
This looks good to me!

TY for the updates on the change requests.

I used the canary you set up in Sandbox for testing and everything worked as expected with the retries.

Good work on this!

nbragdon commented on server/eslintrc.json:
```diff
@@ -0,0 +1,20 @@
+{
+    "env": {
+        "browser": true,
```
I think for the server, `browser` shouldn't be true?

nbragdon commented on server/package.json:
```diff
@@ -5,8 +5,9 @@
+    "start:debug": "node --inspect=0.0.0.0:9229 ./node_modules/.bin/ts-node -r tsconfig-paths/register ./src",
     "start:dev": "nodemon",
-    "test": "nodemon --config ./spec/nodemon.json"
+    "test": "mocha --exit -r ts-node/register spec/*.spec.ts"
```
Looks like some testing changes might have snuck through to this PR

nbragdon reviewed (approved):
Nice job on this, thanks for making the requested changes!
